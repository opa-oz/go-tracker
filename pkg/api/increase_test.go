package api_test

import (
	"github.com/gkampitakis/go-snaps/snaps"
	"net/http"
	"net/http/httptest"
	"testing"
	"tracker/pkg/api"
	"tracker/pkg/middlewares"

	"github.com/gin-gonic/gin"
	"github.com/go-playground/assert/v2"
	"github.com/go-redis/redismock/v9"
)

func TestIncreaseRoute(t *testing.T) {
	r := gin.Default()

	rdb, mock := redismock.NewClientMock()

	r.Use(middlewares.RedisMiddleware(rdb))
	r.GET("/increase/:name", api.Increase)
	r.GET("/i/:name", api.Increase)

	t.Run("if keys is not set, increase returns 1", func(t *testing.T) {
		mock.ExpectGet("test").RedisNil()
		mock.ExpectSet("test", int64(1), 0).SetVal("1")

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/increase/test", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"previous":0,"value":1}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)

		mock.ExpectGet("test").SetVal("1")
		mock.ExpectSet("test", int64(2), 0).SetVal("2")

		w = httptest.NewRecorder()
		req, _ = http.NewRequest(http.MethodGet, "/i/test", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"previous":1,"value":2}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})

	t.Run("if key is set, get returns value", func(t *testing.T) {
		value := "100"
		mock.ExpectGet("test").SetVal(value)
		mock.ExpectSet("test", int64(112), 0).SetVal("112")

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/i/test?step=12", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"previous":100,"value":112}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})

	t.Run("if max is set, get returns less or equal", func(t *testing.T) {
		value := "30"
		mock.ExpectGet("test").SetVal(value)
		mock.ExpectSet("test", int64(50), 0).SetVal("50")

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/increase/test?max=50&step=100", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"previous":30,"value":50}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})
}
