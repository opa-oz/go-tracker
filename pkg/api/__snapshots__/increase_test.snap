
[TestIncreaseRoute/if_keys_is_not_set,_increase_returns_1 - 1]
{"previous":0,"value":1}
---

[TestIncreaseRoute/if_keys_is_not_set,_increase_returns_1 - 2]
{"previous":1,"value":2}
---

[TestIncreaseRoute/if_key_is_set,_get_returns_value - 1]
{"previous":100,"value":112}
---

[TestIncreaseRoute/if_max_is_set,_get_returns_less_or_equal - 1]
{"previous":30,"value":50}
---
